xls2xml/xls2xml/utils.py

```diff
@@ -19,3 +23,135 @@ def header_to_xml_tag(header):
+    :return: False if there is error or True for success
+    :rtype: bool
+    """
+    has_validation_error = False
+    validator = MetadataValidator(validation_schema_filename)
+    keys = reader.get_valid_keys()
+    for key in keys:
+        reader.set_current_key(key)
+        for row in reader:
+            if not validator.validate_data(row, key):
+                has_validation_error = True
+
+    return not has_validation_error
+
+
+def extract_rows(reader, current_key, validation_schema_filename, rows):
+    """
+    Extract data from a XLS worksheet or TSV file named $current_key
+
+    :param reader: reader for the file (XLS or TSV)
+    :type reader: Reader
+    :param current_key: the key name defined in the configuration file
+    :type current_key: basestring
+    :param validation_schema_filename: path to validation rules definition file
+    :type validation_schema_filename:  basestring
+    :param rows: list to append the valid data rows
+    :type rows: list
+    :return: False if there is error or True for success
+    :rtype: bool
+    """
+    reader.set_current_key(current_key)
+    headers = reader.get_current_headers()
+    if not headers:
+        print('There is no header row!', file=sys.stderr)
+        return False
+
+    has_validation_error = False
+    validator = MetadataValidator(validation_schema_filename)
+
+    for row in reader:
+        if validator.validate_data(row, current_key):
+            rows.append(row)
+        else:
+            has_validation_error = True
+
+    if has_validation_error:
+        print("Please fix error first!", file=sys.stderr)
+        return False
+
+    return True
+
+def rows_to_xml(rows, current_key):
+    """
+    Convert data in a list of mappings into an xml tree
+
+    :param rows: a list of mappings
+    :type rows: list
+    :param current_key: a string to indicate which metadata set this is (e.g. Sample)
+    :type current_key: basestring
+    :return: xml tree
+    :rtype: etree.Element
+    """
+    xml = etree.Element(current_key+"Set")
+    for row in rows:
+        element_root = etree.SubElement(xml, current_key)
+        for header in row:
+            child_node = etree.SubElement(element_root, header_to_xml_tag(header))
+            child_node.text = str('' if row.get(header, '') is None else row.get(header, ''))
+    return xml
+
+def transform_xml(input_xml, xslt_filename):
+    """
+    Transform a xml tree to another according to the xslt transformation rules defined in
+    $xslt_filename
+
+    :param input_xml: the input xml tree
+    :type input_xml: etree._Element
+    :param xslt_filename: path to the definition file of xslt transformation rules
+    :type xslt_filename: basestring
+    :return: the transformed xml tree
+    :rtype: etree.Element
+    """
+    xslt_tree = etree.parse(xslt_filename)
+    transform = etree.XSLT(xslt_tree)
+    output_xml = transform(input_xml)
+    return output_xml
+
+def write_to_xml(input_xml, outfile):
+    """
```

Comment:
may be we can name it persist_xml or save_xml_to_disk or file

Reply:
Can't `outfile` be replaced by any other stream handle? If so, the method could be named `write_xml` because it could be written to a file, a string stream, or anything.

Reply:
Renamed to save_xml in 27f748bc